[ui-Side compositing] Support custom scrollbars

nmoucht opened the pull request:
#### b0db30ddd932ac3b92d3f3f814b7f9e22a7bf989
<pre>
[ui-Side compositing] Support custom scrollbars
<a href="https://bugs.webkit.org/show_bug.cgi?id=250926">https://bugs.webkit.org/show_bug.cgi?id=250926</a>
rdar://103890177

Reviewed by Simon Fraser.

Since custom scrollbars are currently handled correctly with UI-side compositing,
we just need to hide the native scrollbars. We do this by making ScrollbarHiddenByStyle
an enum which can now be hidden by custom scrollbars as well. When this gets plumbed to
the UIProcess side, check the ScrollableArea parameters to see if we want to hide the
native scrollbars or not.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::horizontalScrollbarHidden const):
(WebCore::FrameView::verticalScrollbarHidden const):
(WebCore::FrameView::horizontalScrollbarHiddenByStyle const): Deleted.
(WebCore::FrameView::verticalScrollbarHiddenByStyle const): Deleted.
* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setScrollingNodeScrollableAreaGeometry):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
(WebCore::ScrollableAreaParameters::operator== const):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNodeDelegate.h:
(WebCore::ScrollingTreeScrollingNodeDelegate::updateScrollbarsFromStateNode):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarsFromStateNode):
* Source/WebCore/platform/ScrollTypes.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::horizontalScrollbarHidden const):
(WebCore::ScrollableArea::verticalScrollbarHidden const):
(WebCore::ScrollableArea::horizontalScrollbarHiddenByStyle const): Deleted.
(WebCore::ScrollableArea::verticalScrollbarHiddenByStyle const): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::LegacyWebKitScrollingLayerCoordinator::updateScrollingLayer):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::horizontalScrollbarHidden const):
(WebCore::RenderLayerScrollableArea::verticalScrollbarHidden const):
(WebCore::RenderLayerScrollableArea::horizontalScrollbarHiddenByStyle const): Deleted.
(WebCore::RenderLayerScrollableArea::verticalScrollbarHiddenByStyle const): Deleted.
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::commitStateAfterChildren):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::commitStateBeforeChildren):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeOverflowScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeOverflowScrollingNodeRemoteMac::commitStateBeforeChildren):

Canonical link: <a href="https://commits.webkit.org/259322@main">https://commits.webkit.org/259322@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/87dc29e9db9f95f1dd102d8f54dffd136fb82770

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104575 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13654 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37483 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113853 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/174078 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108493 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14771 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4579 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96911 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112805 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110340 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/11398 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/94436 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/38972 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/108052 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/93254 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/26048 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/80636 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/94553 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/7009 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/27406 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/92465 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/4789 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/7127 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/3979 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30049 "Passed tests") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103412 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/13167 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/46962 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/101151 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8916 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25115 "Passed tests") | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3411 "Failed to push commit to Webkit repository") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/ScrollTypes.h:
```diff
@@ -172,6 +172,12 @@ enum class ScrollbarExpansionState : uint8_t {
     Expanded
 };
 
+enum class ScrollbarHidden : uint8_t {
```
I think the code would be more readable if this is more like:
```
enum class NativeScrollbarVisibility : unit8_t {
  Visible,
  HiddenByStyle,
  ReplacedByCustomScrollbar
};
```

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -66,6 +66,22 @@
     ThreadedScrollingTreeScrollingNodeDelegate::updateFromStateNode(scrollingStateNode);
 }
 
+void ScrollingTreeScrollingNodeDelegateMac::updateScrollbarsFromStateNode(const ScrollingStateScrollingNode& scrollingStateNode, CALayer** horizontalLayer, CALayer** verticalLayer)
```
This function doesn't update scrollbars, it just fetches layers, so could use a more appropriate name.

WebKit style is `CALayer **` (note spacing for ObjC types)

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/platform/Scrollbar.cpp:
```diff
@@ -478,4 +478,14 @@ bool Scrollbar::supportsUpdateOnSecondaryThread() const
 #endif
 }
 
+ScrollbarHidden scrollbarHidden(Scrollbar* scrollbar)
```
Make a static function on Scrollbar. Argument can be `const Scrollbar*`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
```diff
@@ -84,8 +84,8 @@ struct ScrollableAreaParameters {
 
-    bool horizontalScrollbarHiddenByStyle { false };
-    bool verticalScrollbarHiddenByStyle { false };
+    NativeScrollbarVisibility horizontalScrollbarHidden { NativeScrollbarVisibility::Visible };
```
I think you should now rename `horizontalScrollbarHidden` and `verticalScrollbarHidden` to match the sense of the enum.

nmoucht replied:
Renamed these and the below functions to horizontalNativeScrollbarVisibility and verticalNativeScrollbarVisibility respectively.

smfr commented on Source/WebCore/platform/ScrollableArea.h:
```diff
@@ -136,8 +136,8 @@ class ScrollableArea : public CanMakeWeakPtr<ScrollableArea>, public CanMakeChec
 
-    virtual bool horizontalScrollbarHiddenByStyle() const { return false; }
-    virtual bool verticalScrollbarHiddenByStyle() const { return false; }
+    virtual NativeScrollbarVisibility horizontalScrollbarHidden() const { return NativeScrollbarVisibility::Visible; }
```
These are now misnamed too.

nmoucht replied:
See above.

smfr commented on Source/WebCore/platform/Scrollbar.h:
```diff
@@ -131,6 +131,7 @@ class Scrollbar : public Widget {
 #endif
 
     bool supportsUpdateOnSecondaryThread() const;
+    static NativeScrollbarVisibility scrollbarHidden(const Scrollbar*);
```
Also misnamed.

nmoucht replied:
Renamed to nativeScrollbarVisibility.

smfr commented on Source/WebCore/rendering/RenderLayerScrollableArea.h:
```diff
@@ -138,8 +138,8 @@ class RenderLayerScrollableArea final : public ScrollableArea {
 
-    bool horizontalScrollbarHiddenByStyle() const final;
-    bool verticalScrollbarHiddenByStyle() const final;
+    NativeScrollbarVisibility horizontalScrollbarHidden() const final;
```
Needs renaming.

nmoucht replied:
See above.

smfr commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -1832,8 +1839,8 @@ header: <WebCore/ScrollingCoordinatorTypes.h>
     bool allowsVerticalScrolling;
-    bool horizontalScrollbarHiddenByStyle;
-    bool verticalScrollbarHiddenByStyle;
+    WebCore::NativeScrollbarVisibility horizontalScrollbarHidden;
```
Need renaming.

nmoucht replied:
See above.

nmoucht committed:
[ui-Side compositing] Support custom scrollbars
https://bugs.webkit.org/show_bug.cgi?id=250926
rdar://103890177

Reviewed by Simon Fraser.

Since custom scrollbars are currently handled correctly with UI-side compositing,
we just need to hide the native scrollbars. We do this by making ScrollbarHiddenByStyle
an enum which can now be hidden by custom scrollbars as well. When this gets plumbed to
the UIProcess side, check the ScrollableArea parameters to see if we want to hide the
native scrollbars or not.

* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::horizontalScrollbarHidden const):
(WebCore::FrameView::verticalScrollbarHidden const):
(WebCore::FrameView::horizontalScrollbarHiddenByStyle const): Deleted.
(WebCore::FrameView::verticalScrollbarHiddenByStyle const): Deleted.
* Source/WebCore/page/FrameView.h:
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::setScrollingNodeScrollableAreaGeometry):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::operator<<):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.h:
(WebCore::ScrollableAreaParameters::operator== const):
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNodeDelegate.h:
(WebCore::ScrollingTreeScrollingNodeDelegate::updateScrollbarsFromStateNode):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.h:
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateScrollbarsFromStateNode):
* Source/WebCore/platform/ScrollTypes.cpp:
(WebCore::operator<<):
* Source/WebCore/platform/ScrollTypes.h:
* Source/WebCore/platform/ScrollableArea.h:
(WebCore::ScrollableArea::horizontalScrollbarHidden const):
(WebCore::ScrollableArea::verticalScrollbarHidden const):
(WebCore::ScrollableArea::horizontalScrollbarHiddenByStyle const): Deleted.
(WebCore::ScrollableArea::verticalScrollbarHiddenByStyle const): Deleted.
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::LegacyWebKitScrollingLayerCoordinator::updateScrollingLayer):
* Source/WebCore/rendering/RenderLayerScrollableArea.cpp:
(WebCore::RenderLayerScrollableArea::horizontalScrollbarHidden const):
(WebCore::RenderLayerScrollableArea::verticalScrollbarHidden const):
(WebCore::RenderLayerScrollableArea::horizontalScrollbarHiddenByStyle const): Deleted.
(WebCore::RenderLayerScrollableArea::verticalScrollbarHiddenByStyle const): Deleted.
* Source/WebCore/rendering/RenderLayerScrollableArea.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::commitStateAfterChildren):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeFrameScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeFrameScrollingNodeRemoteMac::commitStateBeforeChildren):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/ScrollingTreeOverflowScrollingNodeRemoteMac.cpp:
(WebKit::ScrollingTreeOverflowScrollingNodeRemoteMac::commitStateBeforeChildren):

Canonical link: https://commits.webkit.org/259322@main
